### PR TITLE
Add imagePullSecret for CTR

### LINKED ISSUE #4009: Add an "imagePullSecret" property for composedTaskRunner in TaskConfigurationProperties
**Problem description:**
We just added a property for composedTaskRunnerUri in TaskConfigurationProperties. Since this image might be in a private registry it would make sense to have an "imagePullSecret" property for this uri as well.

**Solution description:**
Add an "imagePullSecret" property for composedTaskRunner in TaskConfigurationProperties.

We could restructure the configuration to match:

```
spring:
  cloud:
    dataflow:
      task:
        composedTaskRunner: 
          uri: docker://registry.example.com/spring-cloud-dataflow-composed-task-runner
          imagePullSecret: example-regcred
```

**Description of alternatives:**
It is possible to set "imagePullSecret" for all task apps using:

```
spring:
  cloud:
    dataflow:
      task:
        platform:
          kubernetes:
            accounts:
              default:
                imagePullSecret: example-regcred
```

Since CTR is now part of the distribution and separate from any task apps that are registered it is likely that they are in different registries or use different secrets. It would be better to have a specific entry for the secret to use for the CTR image. 

**Additional context:**

### PR DESCRIPTION
**Pull request by chrisjs**
Resolves #4009

### CODE REVIEW

**Comment by chrisjs** [comment]
below are the manual test cases i went through. i did these for both regular CTR launches and scheduled ones and they seem to work.

pre-reqs:
---
* k8s secret (or however you wanna create it):

```
kubectl create secret docker-registry regcred --docker-server=docker.io --docker-username=YOUR_DOCKER_USER --docker-password=YOUR_DOCKER_PW --docker-email=YOUR_DOCKER@EMAIL
```

* a CTR image in a private repo that gets referenced below in replacement of PRIVATE_REPO

test cases:
---
NOTE: `SPRING_CLOUD_DATAFLOW_TASK_COMPOSEDTASKRUNNER_URI` is currently committed in this PR as `SPRING_CLOUD_DATAFLOW_TASK_COMPOSED_TASK_RUNNER_URI` has been deprecated.

* "new" property namespace in src/kubernetes/server/server-deployment.yaml without image auth

```
        - name: SPRING_CLOUD_DATAFLOW_TASK_COMPOSEDTASKRUNNER_URI
          value: 'docker://springcloud/spring-cloud-dataflow-composed-task-runner:2.7.0-SNAPSHOT'
```

* "old" property namespace in src/kubernetes/server/server-deployment.yaml without image auth:

```
        - name: SPRING_CLOUD_DATAFLOW_TASK_COMPOSED_TASK_RUNNER_URI
          value: 'docker://springcloud/spring-cloud-dataflow-composed-task-runner:2.7.0-SNAPSHOT'
```

* "new" property namespace in src/kubernetes/server/server-deployment.yaml with image auth (need to make secret and private image):

```
        - name: SPRING_CLOUD_DATAFLOW_TASK_COMPOSEDTASKRUNNER_URI
          value: 'docker://PRIVATE_REPO/spring-cloud-dataflow-composed-task-runner:2.7.0-SNAPSHOT'
        - name: SPRING_CLOUD_DATAFLOW_TASK_COMPOSEDTASKRUNNER_IMAGE_PULL_SECRET
          value: 'regcred'

```

now use the CM instead so - remove the above env-vars from from src/kubernetes/server/server-deployment.yaml

* update src/kubernetes/server/server-config.yaml without image auth:

```
    spring:
      cloud:
        dataflow:
          task:
            composedTaskRunner:
              uri: 'docker://springcloud/spring-cloud-dataflow-composed-task-runner:2.7.0-SNAPSHOT'
```

* update src/kubernetes/server/server-config.yaml with image auth:

```
    spring:
      cloud:
        dataflow:
          task:
            composedTaskRunner:
              uri: 'docker://PRIVATE_REPO/spring-cloud-dataflow-composed-task-runner:2.7.0-SNAPSHOT'
              imagePullSecret: regcred
```

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java [hunk 1]
```diff
@@ -16,32 +16,34 @@
...
 
+	@Deprecated
 	public static final String COMPOSED_TASK_RUNNER_NAME = "composed-task-runner";
 
-	/**
-	 * The task application uri to be used for the composed task runner.
-	 */
-	@NotBlank
-	private String composedTaskRunnerUri = "composed-task-runner";
+	@Autowired
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java:45]: In some of our code we make the assumption that composedTaskRunnerConfigurationProperties can be null.    With that assumption should we mark this Autowired as optional.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java [hunk 2]
```diff
@@ -146,4 +155,20 @@ public boolean isUseKubernetesSecretsForDbCredentials() {
...
+
+	private void logDeprecationWarning(String newMethodName) {
+		String callingMethodName = Thread.currentThread().getStackTrace()[2].getMethodName();
+		LOGGER.warn(this.getClass().getName() + "." + callingMethodName
+				+ " is deprecated. Please use " + ComposedTaskRunnerConfigurationProperties.class.getName() + "."
+				+ (newMethodName != null ? newMethodName : callingMethodName));
+	}
+
+	void setComposedTaskRunnerConfigurationProperties(ComposedTaskRunnerConfigurationProperties
+															  composedTaskRunnerConfigurationProperties) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java:171]: May want to reformat this code.    Tabbing is not lining up.   Your friendly neighborhood nitpick.

**Comment by chrisjs** [comment]
updated! the tabs should be just splendid now :100:

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/impl/TaskServiceUtilsTests.java [hunk 3]
```diff
@@ -309,6 +309,48 @@ public void testImagePullSecretNullCTRProperties() {
 				taskDeploymentProperties.containsKey("deployer.composed-task-runner.kubernetes.imagePullSecret"));
 	}
 
+	@Test
+	public void testUseUserAccessTokenFromCTRProps() {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/impl/TaskServiceUtilsTests.java:313]: We should also have a test that  test if the user sets access token to false.   i.e. composedTaskRunnerConfigurationProperties.setUseUserAccessToken(false);

Comment (chrisjs) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/service/impl/TaskServiceUtilsTests.java]: added

**Comment by cppwfs** [comment]
LGTM

**Comment by cppwfs** [comment]
Rebased Squashed merged.

**Pull request closed without merging**